main.go

```diff
@@ -74,12 +74,82 @@ func (s *serialPort) Write(b []byte) (int, error) {
 	return p.Write(b)
 }
 
+var deferredFuncs []func()
+
 func run(ctx context.Context) {
+	// Remove any existing restart file.
+	if err := os.Remove(config.RestartFile()); err != nil && !os.IsNotExist(err) {
+		logger.Errorf("Error removing restart signal file: %v", err)
+	}
+
+	// Setup logging.
+	opts := logger.LogOpts{LoggerName: "OSConfigAgent"}
+	if config.Stdout() {
+		opts.Writers = []io.Writer{os.Stdout}
+	}
+	if runtime.GOOS == "windows" {
+		opts.Writers = append(opts.Writers, &serialPort{"COM1"})
+	}
+
+	// If this call to SetConfig fails (like a metadata error) we can't continue.
+	if err := config.SetConfig(ctx); err != nil {
+		logger.Init(ctx, opts)
```

Comment:
If `SetConfig` fails, won't `logger.Init()` probably fail too? And if not, would it make more sense to call `logger.Init()` before this and include the logic to exit if it fails?

Reply:
Logger.Init wont fail unless there is something very wrong, we need SetConfig to get project id and settings for logger, we only do this here so that we will log somewhere in the event of an issue